arrows/klv/klv_read_write.cxx

```diff
@@ -0,0 +1,170 @@
+
+namespace arrows {
+
+namespace klv {
+
+// ---------------------------------------------------------------------------
+uint64_t
+_imap_pos_infinity( size_t length )
+{
+  return length ? 0xC8ull << ( ( length - 1 ) * 8 ) : 0;
+}
+
+// ---------------------------------------------------------------------------
+uint64_t
+_imap_neg_infinity( size_t length )
+{
+  return length ? 0xE8ull << ( ( length - 1 ) * 8 ) : 0;
+}
+
+// ---------------------------------------------------------------------------
+uint64_t
+_imap_pos_quiet_nan( size_t length )
+{
+  return length ? 0xD0ull << ( ( length - 1 ) * 8 ) : 0;
+}
+
+// ---------------------------------------------------------------------------
+uint64_t
+_imap_neg_quiet_nan( size_t length )
+{
+  return length ? 0xF0ull << ( ( length - 1 ) * 8 ) : 0;
+}
+
+// ---------------------------------------------------------------------------
+uint64_t
+_imap_pos_signal_nan( size_t length )
+{
+  return length ? 0xD8ull << ( ( length - 1 ) * 8 ) : 0;
+}
+
+// ---------------------------------------------------------------------------
+uint64_t
+_imap_neg_signal_nan( size_t length )
+{
+  return length ? 0xF8ull << ( ( length - 1 ) * 8 ) : 0;
+}
+
+// ---------------------------------------------------------------------------
+_imap_terms
+_calculate_imap_terms( double minimum, double maximum, size_t length )
+{
+  // ST1201, Section 8.1.2
+  auto const float_exponent = std::ceil( std::log2( maximum - minimum ) );
+  auto const int_exponent = 8.0 * length - 1.0;
+
+  _imap_terms result = {};
+  result.forward_scale = std::exp2( int_exponent - float_exponent );
+  result.backward_scale = std::exp2( float_exponent - int_exponent );
+  result.zero_offset = ( minimum < 0 && maximum > 0 )
+                       ? result.forward_scale * minimum -
+                       std::floor( result.forward_scale * minimum )
+                       : 0.0;
+  return result;
+}
+
+// ---------------------------------------------------------------------------
+size_t
+klv_imap_length( double minimum, double maximum, double precision )
+{
+  // ST1201, Section 8.1.1
+  _check_range_precision( minimum, maximum, precision );
+
+  auto const length_bits = std::ceil( std::log2( maximum - minimum ) ) -
+                           std::floor( std::log2( precision ) ) + 1;
+  return static_cast< size_t >( std::ceil( length_bits / 8.0 ) );
+}
+
+// ---------------------------------------------------------------------------
+double
+klv_imap_precision( double minimum, double maximum, size_t length )
+{
+  _check_range_length( minimum, maximum, length );
+
+  auto const length_bits = length * 8.0;
+  return std::exp2( std::log2( maximum - minimum ) - length_bits + 1 );
+}
+
+namespace {
+
+// ---------------------------------------------------------------------------
+void
+_check_range( double minimum, double maximum )
+{
+  if( !std::isfinite( minimum ) )
+  {
+    VITAL_THROW( kv::invalid_value, "minimum must be finite" );
```

Comment:
While I get wanting to use a KWIVER exception, these feel suspiciously like `std::logic_error`s. (Do the min/max ever come from external inputs? Or are they likely to be hard-coded in the source code of whoever uses these functions?)

Same for `_check_range_precision`...